### PR TITLE
Adding layers to the explainer

### PR DESCRIPTION
**Pull request by Juanmihd**
This PR adds the explainer for layers, it also does some refactor to the readme, to include a section of currently released changes on the spec.

### CODE REVIEW

File: spec/layers.md [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+
+Currently, the only way to draw a series of drawing calls at the same time, while applying the same sort of effects (shadows, blur, alpha, filters, compositing) to all of the drawn objects, without having these effects being applied one to one to this object, is by using an auxiliary canvas and then, drawing that auxiliary canvas to the canvas.
+This proposal adds an alternative to that (the same way that Fullter in Google and Core Graphics at Apple do) by creating a layer that then will be drawn in one go when the layer ends.
+
+
+Proposal
+--------
+
+```webidl
+interface CanvasRenderingContext2D {
```
Comment (fserb) [inline spec/layers.md:19]: should this be part of `CanvasState`?

File: spec/layers.md [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+};
+
+interface OffscreenCanvasRenderingContext2D {
+  // extending original
+  void beginLayer();
+  void endLayer();
+};
+```
+
+`beginLayer()` sets the start point of the layer, it will also specify the effects (shadows, blur, alpha, filters, compositing) that will be used to draw the layer.
```
Comment (fserb) [inline spec/layers.md:32]: maybe rephrase this by saying that the current state of the canvas specifies the effects that will be used. And list specifically  which part of the state is considered and which part is not.

Also, that we "reset" that part of the state inside the layer.

File: spec/layers.md [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+`endLayer()` sets the end point of the layer. At that moment the layer itself will be drawn as one single object into the canvas. When the layer ends, the
```
Comment (fserb) [inline spec/layers.md:33]: missing something in the end here.

File: spec/layers.md [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+Example usage
+-------------
+
+```js
+// Javascript example
+const canvas = document.createElement('canvas');
+const ctx = canvas.getContext('2d');
+
+ctx.beingLayer();
+ctx.globalAlpha = 0.5;
```
Comment (fserb) [inline spec/layers.md:45]: the globalAlpha should come before, no?

File: spec/layers.md [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+ctx2.fillRect(70,70,75,50);
+
+ctx.globalAlpha = 0.5; 
+ctx.drawImage(canvas2,0,0);
+```
+
+Alternatives considered
+-----------------------
+
+None.
```
Comment (fserb) [inline spec/layers.md:75]: we considered adding the properties of the layer as parameters to `beginLayer()`. We gave up on this because it created a new semantic for the 2D API.

File: spec/layers.md [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+
+
+References
+----------
+
+None.
```
Comment (fserb) [inline spec/layers.md:81]: maybe link to the relevant APIs on CG, Skia, etc?

File: spec/layers.md (hunk 1 above)
Comment (Juanmihd) [reply spec/layers.md:19]: I think these methods are part of the Context itself, the same way as save and restore.

**Comment by Juanmihd** [comment]
Fernando, I addressed the comments, could you take another look? Thanks!

File: spec/layers.md [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+  void endLayer();
+};
+
+interface OffscreenCanvasRenderingContext2D {
+  // extending original
+  void beginLayer();
+  void endLayer();
+};
+```
+The rendering of the layer will use the current state of the canvas, and will reset these effects to the defaults at the end of the layer. The effects that will be used are:
```
Comment (Kaiido) [inline spec/layers.md:31]: That sounds impractical.

If these values are really reset to their default at `endLayer()`, how come we can have nested layers? How do you do gCO=source-in inside a gCO=lighter inside a globalAlpha=0.2?  Aren't the gCO=lighter and globalAlpha=0.2 reset by the first inner layer?
Not sure yet what would be the best ergonomics, but should we just take `endLayer()` to mean "render this layer as would drawImage do", i.e using all the context's properties at the time of the call to `endLayer()`? On this case maybe we could bikeshed over `endLayer()` vs `paintLayer()` & al.

File: spec/layers.md [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+- filter
+
+`beginLayer()` sets the start point of the layer, it also captures the current state of the canvas (see list above) that will be used when rendering the layer.
+`endLayer()` sets the end point of the layer. At that moment the layer itself will be drawn as one single object into the canvas. When the layer ends, we reset all the effects used in the drawing of the layer at the end of it (see list above).
+
+These methods are nesteable, so layers can be created and drawn within layers.
+
+If there is a dangling `beginLayer()` without `endLayer()`, those operations will not be drawn.
+
+For the interactions with `save()` and `restore()`, a `save()` without `restore()` (or single `restore()` without `save()`) between a pair of `beginLayer()` and `endLayer()` will be discarded.
```
Comment (Kaiido) [inline spec/layers.md:47]: Is this necessary? It feels like that will mix badly with `clip()` (which still can't be discarded without save/restore).

Wouldn't it be easier to state that layers are not saved by `save()`, and that they keep doing what they are doing?

File: spec/layers.md [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+Example usage
+-------------
+
+```js
+// Javascript example
+const canvas = document.createElement('canvas');
+const ctx = canvas.getContext('2d');
+
+ctx.globalAlpha = 0.5; 
+ctx.beingLayer();
```
Comment (Kaiido) [inline spec/layers.md:58]: Typo: beginLayer

**Review by Kaiido — COMMENTED** [review]
Funny this comes now, we [discussed about this issue](https://github.com/whatwg/html/issues/5621#issuecomment-956652873) not more than 2 days ago, and I am glad to see this proposal coming to life.  
Thank you for this.

I only have a few points of concerns though:

File: spec/layers.md (hunk 1 above)
Comment (Kaiido) [reply spec/layers.md:19]: I think @fserb is right here, `save()` and `restore()` are defined on [CanvasState](https://html.spec.whatwg.org/multipage/canvas.html#canvasstate). This allows other interfaces like OffscreenCanvasRenderingContext2D or PaintWorklet's context to also inherit it, without having the full pack.

File: spec/layers.md [hunk 10]
```diff
@@ -0,0 +1,97 @@
...
+--------
+
+```webidl
+interface mixin CanvasState {
+  // extending original
+  undefined beginLayer();
+  undefined endLayer();
+};
+```
+The rendering state used for the layer rendering will use the current state of the canvas. The effects that will be used are:
```
Comment (fserb) [inline spec/layers.md:25]: maybe rephrase this, as they are not "effects", and move it after the beginLayer. paragraph.

File: spec/layers.md (hunk 7 above)
Comment (Kaiido) [reply spec/layers.md]: I think I get what doesn't work here, you probably meant to say that `endLayer()` will reset the properties to the ones saved by the latest `beginLayer()` and not "to the defaults" (here and below)

File: spec/layers.md (hunk 1 above)
Comment (Juanmihd) [reply spec/layers.md]: Done!

**Comment by fserb** [comment]
I'm merging this as is, but I want to keep discussing it, @Kaiido and see what we can do better.

File: spec/layers.md (hunk 7 above)
Comment (Juanmihd) [reply spec/layers.md]: Hi Kaiido, I'm sorry, when you reviewed this explainer, it was still in WIP, and I had some things wrong. Could you take a new look now? Thanks!

**Comment by Juanmihd** [comment]
Hi @Kaiido, I'm sorry, when you reviewed this explainer, it was still in WIP, and I had some things wrong. Could you take a new look now? Thanks!

**Comment by Kaiido** [comment]
The new commit did answer all my questions, much clearer thanks, and I didn't spot any more typos in there ;-)  

I will open an issue regarding the idea of a CanvasFilter model later this week, but I actually see that it would be about the same as the [CanvasRecording](https://github.com/fserb/canvas2D/blob/master/spec/recording.md) proposal, that I somehow missed all this time, except that the `renderLayer()` would be more like a `drawImage()`, using whatever composition status the context is in for rendering. But once again, I'll try to develop this in its own issue during the week.
